app/src/main/java/com/kamron/pogoiv/MainActivity.java

```diff
@@ -295,13 +295,42 @@ private void startPokeFly() {
         Intent intent = Pokefly.createIntent(this, trainerLevel, statusBarHeight, batterySaver);
         startService(intent);
 
-        pokeFlyRunning = true;
-
         if (settings.shouldLaunchPokemonGo()) {
             openPokemonGoApp();
         }
     }
 
+    private boolean isPokeFlyRunning() {
+        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
+        for (ActivityManager.RunningServiceInfo serviceInfo : manager.getRunningServices(Integer.MAX_VALUE)) {
+            if (Pokefly.class.getName().equals(serviceInfo.service.getClassName())) {
+                return true;
+            }
+        }
```

Comment:
Now on a PC, I can try a longer comment.

If I'm not mistaken, the existing code uses flags in MainActivity that can easily get out of sync with Pokefly in case it's killed for memory, etc. With notification actions to stop Pokefly, it WILL get out of sync.

The commit above tries to solve this by looking at all the services running on the system. This is very expensive in terms of CPU and memory usage and doesn't really solve the problem completely.

What if user in the MainActivity screen, but uses notification to stop the service and then presses the button in the MainActivity to trigger this MainActivity code. If done fast enough (and threads get context switched out), it can cause the function to return a wrong value. The absolute correct code would need locking and synchronization I think or some kind of message queue to serialize all actions (Android might be doing this for us depending on how many threads it creates per task -- I don't have a confident answer).

Anyway, my suggestion was something like this in Pokefly. You don't have to use it and it's not significantly more race-condition proof than your implementation, but I think it'll be just as correct as the above implementation and less resource hungry.

```
public class Pokefly extends Service {

private static boolean running = false;

onCreate() {
  running = true; // Say running while you are starting.
  // do init stuff
}

onDestroy() {
  // do exit stuff
  running = false; // Say running as long as you haven't stopped.
}

public static boolean isRunning() {
  return running;
}
```